src/gstring.rs

```diff
@@ -33,11 +33,37 @@ unsafe impl Send for GString {}
 unsafe impl Sync for GString {}
 
 impl GString {
+    /// Create a new GString from a glib-originated string, taking ownership.
+    ///
+    /// # Safety
+    ///
+    /// The provided string must be a valid C string (i.e. `'0'` terminated).
+    ///
+    /// The provided pointer must be allocated by glib such that `g_free(ptr)`
+    /// is valid, as the `GString` will call `g_free(ptr)` on drop.
+    ///
+    /// It is essential that noone else free this pointer as it owned by the
+    /// returned `GString`.
+    ///
+    /// The underlying string must not be mutated, in particular in terms of
+    /// length, underneath the `GString` instance.
     pub unsafe fn new(ptr: *mut c_char) -> Self {
         assert!(!ptr.is_null());
         GString(Inner::Foreign(ptr, libc::strlen(ptr)))
     }
 
+    /// Create a new GString from a glib-originated string, borrowing it rather
+    /// than taking ownership.
+    ///
+    /// # Safety
+    ///
+    /// The provided string must be a valid C string (i.e. `'0'` terminated).
+    ///
+    /// It is essential that noone else free this pointer as it owned by the
+    /// returned `GString` until the borrow is dropped.
+    ///
+    /// The underlying string must not be mutated, in particular in terms of
+    /// length, underneath the `GString` instance for the duration of the borrow.
     pub unsafe fn new_borrowed(ptr: *const c_char) -> Borrowed<Self> {
```

Comment:
Why do we have this and the other constructor at all? IMHO they should be removed, the trait impls are enough and what is used everywhere else.

@GuillaumeGomez What do you think?

Reply:
It's not supposed to be handled by users so it's strange to have them. Why did we add these functions in the first place? To not use `from_glib`?

Reply:
I think just an oversight. Let's remove them then, @kinnison :)

Reply:
If I delete them and nothing goes wrong in the CI then I guess it's okay, though we should remember to boost the minor to 0.10 on the next release since this is incompatible.

Reply:
To remove them I'd need to shunt stuff around a lot.  I could do that, but these are used in many traits.  Perhaps better would be to simply make the `pub(crate)` or even private since they're only used inside `gstring.rs` ?

Reply:
That's going to happen anyway, we already had many other incompatible changes. The reason why the version is still at 0.9 in git is something @GuillaumeGomez has to explain to you (I disagree and would've bumped it to 0.10 long ago)

Reply:
Private also works for me, just not public :)

Reply:
I think private is fine. I've added a commit to that effect.